src/c++/perf_analyzer/genai-perf/genai_perf/llm_inputs/synthetic_prompt_generator.py

```diff
@@ -65,28 +65,54 @@ def _create_farewell_lines(cls) -> List[str]:
         return farewell_lines
 
     @classmethod
-    def _create_prompt_from_farewell_lines(
+    def _create_prompt_from_lines(
         cls,
-        remaining_prompt_tokens: int,
-        farewell_lines: List[str],
+        requested_prompt_tokens: int,
+        source_lines: List[str],
         tokenizer: Tokenizer,
     ) -> str:
-        prompt = ""
         get_token_length = lambda text: len(tokenizer.encode(text))
 
-        sampling_lines = True
-        while sampling_lines:
-            for line in farewell_lines:
-                line_to_add = line
-                if remaining_prompt_tokens - get_token_length(line_to_add) < 0:
-                    # This will cut off a line in the middle of a word, but that's ok since an
-                    # llm should be able to handle that.
-                    line_to_add = line_to_add[: int(math.ceil(remaining_prompt_tokens))]
-                    sampling_lines = False
-                    prompt += line_to_add
-                    break
-                prompt += line_to_add
-                remaining_prompt_tokens -= get_token_length(line_to_add)
+        line_iterator = itertools.cycle(source_lines)
+
+        def word_generator():
+            while True:
+                next_line = next(line_iterator)
+                words = re.split("[ \n]+", next_line)
+                for word in words:
+                    yield word
+
+        word_iterator = word_generator()
+
+        # Fast add lines
+        remaining_tokens = requested_prompt_tokens
+        prompt = ""
+        num_tokens_in_avg_line = get_token_length(source_lines[0] + source_lines[1]) / 2
+        num_lines_to_add_fast = math.floor(
+            0.5 * requested_prompt_tokens / num_tokens_in_avg_line
+        )
+        while num_lines_to_add_fast:
+            for _ in range(num_lines_to_add_fast):
+                next_line = next(line_iterator)
+                prompt = prompt + next_line
+
+            curr_tokens = get_token_length(prompt)
+            remaining_tokens = requested_prompt_tokens - curr_tokens
+            num_lines_to_add_fast = math.floor(
+                0.5 * remaining_tokens / num_tokens_in_avg_line
+            )
```

Comment:
Maybe this is more for the refactor, but I had trouble following what was going on here.

This is only run once at the beginning of the call to genai-perf, right?

Could this be rewritten in a slower, but more readable way? And then if it's slow enough to hurt UX, we can optimize?

Reply:
I had a similar thought but this PR is supposed to be a quick fix to be cherry-picked, so making it readable is likely a followup ticket.